File: pari3h.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* reverseList(struct Node* head) {
    struct Node* prev = NULL;
    struct Node* current = head;
    struct Node* next = NULL;
    while (current != NULL) {
        next = current->next; 
        current->next = prev; 
        prev = current;       
        current = next;
    }
    head = prev;
    return head;
}

void printList(struct Node* head);
struct Node* createListFromArray(int arr[], int n);

int main() {
    int arr[] = {1, 2, 3, 4, 5};
    struct Node* head = createListFromArray(arr, 5);
    
    printf("Original list: ");
    printList(head);
    
    head = reverseList(head);
    
    printf("Reversed list: ");
    printList(head);

    return 0;
}

File: pari3h.py
def reverse_list(self):
    prev = None
    current = self.head
    while current:
        next_node = current.next
        current.next = prev
        prev = current
        current = next_node
    self.head = prev

llist = SinglyLinkedList()
llist.insert_at_end(1)
llist.insert_at_end(2)
llist.insert_at_end(3)
llist.insert_at_end(4)
print("Original Linked List:")
llist.print_list()

llist.reverse_list()
print("Reversed Linked List:")
llist.print_list()
